### PR TITLE
Create wa-tor algorithm

### PR DESCRIPTION
**Pull request by CaedenPH**
### Describe your change:
Implements the wa-tor algorithm (https://en.wikipedia.org/wiki/Wa-Tor)

The wa-tor algorithm is a cellular automata featuring predators and prey. All code is doctested and documented, linking to multiple sources.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by CaedenPH** [comment]
@cclauss Large file to review, apologies for that. All functions are doctests and documented, with comments explaining all the code.
Build is failing because of another file, see #8900

File: cellular_automata/wa_tor.py [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+        >>> e = Entity(False, coords=(0, 0))
+        >>> e.reset_reproduction_time()
+        >>> e.remaining_reproduction_time == PREDATOR_REPRODUCTION_TIME
+        True
+        """
+        self.remaining_reproduction_time = (
+            PREY_REPRODUCTION_TIME if self.prey is True else PREDATOR_REPRODUCTION_TIME
+        )
+
+    def __repr__(self) -> str:
```
Comment (cclauss) [inline cellular_automata/wa_tor.py:77]: [`object.__repr__()`](https://docs.python.org/3/reference/datamodel.html#object.__repr__)
> If at all possible, this should look like a valid Python expression that could be used to recreate an object with the same value (given an appropriate environment)

File: cellular_automata/wa_tor.py [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+
+    time_passed: Callable[["WaTor", int], None] | None
+
+    def __init__(self, width: int, height: int) -> None:
+        self.width = width
+        self.height = height
+        self.time_passed = None
+
+        self.planet: list[list[Entity | None]] = [
+            [None for _ in range(width)] for _ in range(height)
```
Comment (cclauss) [inline cellular_automata/wa_tor.py:124]: Is this faster er not?
```suggestion
            [None] * width for _ in range(height)
```

Comment (CaedenPH) [reply cellular_automata/wa_tor.py:124]: Probably a negligible difference, but its less code so I'll merge

File: cellular_automata/wa_tor.py (hunk 1 above)
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:77]: For some reason I correlate repr with <a=1 b=2> objects

**Comment by CaedenPH** [comment]
Again, not sure why build is failing

**Comment by tianyizheng02** [comment]
```
Fatal Python error: Illegal instruction

Thread 0x00007fae8081e640 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/concurrent/futures/thread.py", line 81 in _worker
  File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/threading.py", line 975 in run
  File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/threading.py", line 1038 in _bootstrap_inner
  File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/threading.py", line 995 in _bootstrap

Current thread 0x00007faf07b3cb80 (most recent call first):
  File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/site-packages/projectq/backends/_sim/_simulator.py", line 83 in __init__
  File "/opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/site-packages/projectq/cengines/_main.py", line 135 in __init__
  File "<doctest quantum.quantum_random.get_random_number[0]>", line 1 in <module>
  ...

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg,
...
/home/runner/work/_temp/96683403-d4c6-40d5-997a-4fb01d92a187.sh: line 1:  1824 Illegal instruction     (core dumped) pytest --ignore=quantum/q_fourier_transform.py --ignore=project_euler/ --ignore=scripts/validate_solutions.py --cov-report=term-missing:skip-covered --cov=. .
quantum/quantum_random.py 
Error: Process completed with exit code 132.
```

It looks to be something in one of `quantum/quantum_random.py`'s doctests causing a segfault. The automatic builds on my fork have been failing because of this too.

**Comment by cclauss** [comment]
Please rename `quantum/quantum_random.py` —> `quantum/quantum_random.py.DISABLED`

**Comment by tianyizheng02** [comment]
@CaedenPH Let's see if the build succeeds now

**Comment by CaedenPH** [comment]
Works 👍

**Comment by cclauss** [comment]
@tianyizheng02 Can you please approve and merge if you think this is ready and useful?

File: cellular_automata/wa_tor.py [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+            # Balance out the predators and prey
+            self.balance_predators_and_prey()
+
+            if self.time_passed is not None:
+                # Call time_passed function for Wa-Tor planet
+                # visualisation in a terminal or a graph.
+                self.time_passed(self, iter_num)
+
+
+def display_visually(wt: WaTor, iter_number: int, *, colour: bool = True) -> None:
```
Comment (tianyizheng02) [inline cellular_automata/wa_tor.py:518]:
```suggestion
def visualize(wt: WaTor, iter_number: int, *, colour: bool = True) -> None:
```

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]
General feedback:
1) Per PEP8, `x is True` -> `x`
2) Rather than iterating over the 4 possible directions and having multiple nested if-statements handling each direction, it's a lot simpler to just iterate over a list of the 4 neighboring squares and handle them in a more generic manner. For example, instead of checking if `row - 1 >= 0` when the direction is North and if `row + 1 < self.height` when the direction is South, just check whether the row to move to is in bounds: `0 <= r < self.height`. An extra comparison or two is worth the increased readability, IMO.

File: cellular_automata/wa_tor.py [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+    ansii_colour_end = "\x1b[0m " if colour is True else " "
+
+    planet = wt.planet
+    output = ""
+
+    # Iterate over every entity in the planet
+    for i in range(len(planet)):
+        for j in range(len(planet[0])):
+            if (entity := planet[i][j]) is None:
+                output += " . "
```
[the remark this answers is not in the document]
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:557]: Good spot

File: cellular_automata/wa_tor.py [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+    .  #  .
+    <BLANKLINE>
+    Iteration: 0 | Prey count: 2 | Predator count: 3 |
+    """
+    if colour is True:
+        __import__("os").system("")
+        print("\x1b[0;0H\x1b[2J\x1b[?25l")
+
+    reprint = "\x1b[0;0H" if colour is True else ""
+    ansii_colour_end = "\x1b[0m " if colour is True else " "
```
[the remark this answers is not in the document]
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:548]: Yep, typo

File: cellular_automata/wa_tor.py [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+                        "\x1b[38;2;96;241;151m"
+                        if entity.prey
+                        else "\x1b[38;2;255;255;15m"
+                    )
+                output += f" {'#' if entity.prey else 'x'}{ansii_colour_end}"
+
+        output += "\n"
+
+    entities = wt.get_entities()
+    prey_count = len(list(filter(lambda entity: entity.prey is True, entities)))
```
[the remark this answers is not in the document]
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:570]: > ```suggestion
> 
>     prey_count = sum(entity.prey for entity in entities)
> 
> ```
> 
> `True` is interpreted as 1 and `False` is interpreted as 0, so you can just count the number of `True`s using `sum`

Good spot

File: cellular_automata/wa_tor.py [hunk 7]
```diff
@@ -0,0 +1,587 @@
...
+        >>> wt.balance_predators_and_prey()
+        >>> len(wt.get_entities()) == entities
+        False
+        """
+        entities = self.get_entities()
+        shuffle(entities)
+
+        if len(entities) >= MAX_ENTITIES - MAX_ENTITIES / 10:
+            prey = list(filter(lambda entity: entity.prey is True, entities))
+            predators = list(filter(lambda entity: entity.prey is True, entities))
```
[the remark this answers is not in the document]
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:207]: Correct, good catch

File: cellular_automata/wa_tor.py [hunk 8]
```diff
@@ -0,0 +1,587 @@
...
+            prey_count, predator_count = len(prey), len(predators)
+
+            if prey_count > predator_count:
+                for entity in prey[:DELETE_UNBALANCED_ENTITIES]:
+                    # Purge the first n entities of the prey
+                    self.planet[entity.coords[0]][entity.coords[1]] = None
+            else:
+                for entity in predators[:DELETE_UNBALANCED_ENTITIES]:
+                    # Purge the first n entities of the predators
+                    self.planet[entity.coords[0]][entity.coords[1]] = None
```
[the remark this answers is not in the document]
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:218]: Nice

File: cellular_automata/wa_tor.py (hunk 3 above)
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:517]: I'll make it visualise;

File: cellular_automata/wa_tor.py [hunk 9]
```diff
@@ -0,0 +1,587 @@
...
+
+        >>> wt = WaTor(WIDTH, HEIGHT)
+        >>> len(wt.get_entities()) == PREDATOR_INITIAL_COUNT + PREY_INITIAL_COUNT
+        True
+        """
+        start: Any = []
+        return sum(
+            [[entity for entity in column if entity] for column in self.planet],
+            start=start,
+        )
```
[the remark this answers is not in the document]
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:185]: I agree that using multiple fors is better think this syntax looks wrong to me, I'll check it out

File: cellular_automata/wa_tor.py [hunk 10]
```diff
@@ -0,0 +1,587 @@
...
+            if prey_count > predator_count:
+                for entity in prey[:DELETE_UNBALANCED_ENTITIES]:
+                    # Purge the first n entities of the prey
+                    self.planet[entity.coords[0]][entity.coords[1]] = None
+            else:
+                for entity in predators[:DELETE_UNBALANCED_ENTITIES]:
+                    # Purge the first n entities of the predators
+                    self.planet[entity.coords[0]][entity.coords[1]] = None
+
+    def get_surrounding_prey(self, entity: Entity) -> list[Entity]:
```
[the remark this answers is not in the document]
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:215]: This doesn't weight the locations based off of the weighting parameters

[the remark this answers is not in the document]
Comment (tianyizheng02) [reply cellular_automata/wa_tor.py:215]: What weighting parameters are you referring to? I don't see `get_surrounding_prey` using any weight values

[the remark this answers is not in the document]
Comment (CaedenPH) [reply cellular_automata/wa_tor.py:215]: @tianyizheng02 You're absolutely right for this function, no weighting and can definitely use that. Will also work in some logic for the other review where I generate the adjacent list and then shuffle accordingly

**Review by tianyizheng02 — APPROVED** [review]
LGTM

**Pull request merged**